SQL Store.adoc

```diff
@@ -0,0 +1,256 @@
+* a protobuf schema can either be generated dynamically from the resultset metadata or supplied by the user
+* the key can either be a single column or composite
+* the store is to be implemented as part of the existing JDBC store module so that it can reuse the general configuration of connection pooling/datasources and the dialect helpers
+
+== Modes
+
+=== Table mode
+
+This mode is the simplest, in that it just requires the user to supply a single table name.
+Through the database metadata, the store will determine:
+
+* the column names and types
+* which columns are part of the primary key
+
+The proposed XML configuration for this mode:
+
+[source,xml]
+----
+<local-cache name="books">
+   <persistence>
+      <table-jdbc-store>
+         <data-source jndi-url="..." />
+         <!-- the table element is required -->
+         <table name="books"/>
+         <!-- the schema element is optional. Defaults will be used otherwise -->
+         <schema name="Book" key-name="BookID" package="library"/>
+      </table-jdbc-store>
+   </persistence>
+</local-cache>
+----
+
+=== Query mode
+
+In this mode, the user supplies the SQL queries to retrieve the entries and (optionally) to insert/update/delete them.
+
+[cols="1,2,1"]
+|===
+| *Query*
+| *Description*
+| *Example*
+
+| *SELECT ALL*
+| The query to use when bulk-loading data into the cache. Columns *MUST* be explicitly listed (wildcards are *NOT* allowed). Expressions are allowed but they *MUST* be labelled.
+| `select isbn, title from books`
+
+| *SELECT SINGLE*
+| The query to use when loading a single entry into the cache. Columns *MUST* be explicitly listed (wildcards are *NOT* allowed). Expressions are allowed but they *MUST* be labelled.
+| `select isbn, title from books where isin = :key`
+
+| *UPSERT* 
+| The query to use when inserting/updating entrys into the database.
+| `insert into books (isbn, title) values() ON DUPLICATE KEY UPDATE`
+
+| *DELETE*
+| The query to use when a single entry is deleted.
+| `delete from books where id = :key`
+
+| *DELETE ALL*
+| The query to use when the cache is cleared.
+| `delete from books/truncate table books`
+
+|===
+
+The proposed XML configuration for this mode:
+
+[source,xml]
+----
+<local-cache name="books">
+   <persistence>
+      <query-jdbc-store>
+         <data-source jndi-url="..." />
+         <!-- the queries element is required, including the names of all the columns which make up the key -->
+         <queries key-columns="isbn">
+            <select-all>SELECT isbn, title FROM books</select-all>
+            <select-single>SELECT isbn, title FROM books WHERE isbn = :key</select-single>
+            <delete-all>DELETE FROM books</delete-all>
+            <delete>DELETE FROM books WHERE isbn = :key</delete>
+            <upsert>INSERT INTO books (isbn, title) VALUES (:key, :value) ON CONFLICT (isbn) DO UPDATE SET title = :value</upsert>
+         </queries>
```

Comment:
What about a size query?

Reply:
Good idea.

Reply:
@rvansa added

Reply:
@tristantarrant I also wonder, since the recent updates to the stores have been quite segment-focused, shouldn't there be a possibility to keep the segment ids around?

That's not in sync with the initial story on the top (that sounds like a read-only story but then you expand into a full-fledged configurable sql store) where you start with existing DB - but if the is supposed to be *The JDBC store* it could fit. But then you might start thinking about metadata...

Reply:
I shied away from metadata for this iteration, but that would be the next evolutionary step. This would require a separate table with a `FOREIGN KEY` and `ON DELETE CASCADE`.

Reply:
Understood. I guess the segment id comes in a next step, too, as it requires a new set of operations anyway.